src/GafferSceneUI/LightTool.cpp

```diff
@@ -0,0 +1,1916 @@
+		// Angles are in "handle-space" (generally 1/2 the full cone for the cone angle
+		// and the full penumbra angle for penumbras. See `handleAngles` and `plugAngles`
+		// for conversion details.)
+		struct DragStartData
+		{
+			Inspector::ResultPtr coneInspection;
+			float originalConeHandleAngle;
+			Inspector::ResultPtr penumbraInspection;
+			std::optional<float> originalPenumbraHandleAngle;
+		};
+
+	public :
+
+		enum class HandleType
+		{
+			Cone,
+			Penumbra
+		};
+
+		SpotLightHandle(
+			const std::string &attributePattern,
+			HandleType handleType,
+			SceneViewPtr view,
+			const float zRotation,
+			const std::string &name = "SpotLightHandle"
+		) :
+			LightToolHandle( attributePattern, name ),
+			m_view( view ),
+			m_zRotation( zRotation ),
+			m_handleType( handleType ),
+			m_visualiserScale( 1.f ),
+			m_frustumScale( 1.f ),
+			m_lensRadius( 0 ),
+			m_dragStartData {}
+		{
+
+			mouseMoveSignal().connect( boost::bind( &SpotLightHandle::mouseMove, this, ::_2 ) );
+
+		}
+		~SpotLightHandle() override
+		{
+
+		}
+
+		void update( ScenePathPtr scenePath, const PlugPtr &editScope ) override
+		{
+			LightToolHandle::update( scenePath, editScope );
+
+			if( !handleScenePath()->isValid() )
+			{
+				m_coneAngleInspector.reset();
+				m_penumbraAngleInspector.reset();
+				return;
+			}
+
+			ConstCompoundObjectPtr attributes = handleScenePath()->getScene()->fullAttributes( handleScenePath()->names() );
+
+			float defaultVisualiserScale = 1.f;
+			if( auto p = m_view->descendant<const FloatPlug>( "drawingMode.visualiser.scale" ) )
+			{
+				defaultVisualiserScale = p->getValue();
+			}
+			auto visualiserScaleData = attributes->member<FloatData>( g_lightVisualiserScaleAttributeName );
+			m_visualiserScale = visualiserScaleData ? visualiserScaleData->readable() : defaultVisualiserScale;
+
+			float defaultFrustumScale = 1.f;
+			if( auto p = m_view->descendant<const FloatPlug>( "drawingMode.light.frustumScale" ) )
+			{
+				defaultFrustumScale = p->getValue();
+			}
+			auto frustumScaleData = attributes->member<FloatData>( g_frustumScaleAttributeName );
+			m_frustumScale = frustumScaleData ? frustumScaleData->readable() : defaultFrustumScale;
+
+			/// \todo This can be simplified and some of the logic, especially getting the inspectors, can
+			/// be moved to the constructor when we standardize on a single USDLux light representation.
+
+			for( const auto &[attributeName, value] : attributes->members() )
+			{
+				if(
+					StringAlgo::match( attributeName, attributePattern() ) &&
+					value->typeId() == (IECore::TypeId)ShaderNetworkTypeId
+				)
+				{
+					const auto shader = attributes->member<ShaderNetwork>( attributeName )->outputShader();
+					std::string shaderAttribute = shader->getType() + ":" + shader->getName();
+
+					auto coneParameterName = Metadata::value<StringData>( shaderAttribute, "coneAngleParameter" );
+					assert( coneParameterName );
```

Comment:
This assert fails/a release build will crash in the following situation:

- Activate the light tool
- Select any light that isn't a spotLight

```
gaffer: src/GafferSceneUI/LightTool.cpp:646: virtual void {anonymous}::SpotLightHandle::update(GafferScene::ScenePathPtr, const PlugPtr&): Assertion `coneParameterName' failed.

__assert_fail () from /lib64/libc.so.6
(anonymous namespace)::SpotLightHandle::update (this=0xbf64fc0, scenePath=..., editScope=...) at src/GafferSceneUI/LightTool.cpp:646
GafferSceneUI::LightTool::updateHandleInspections (this=0xcb77a10) at src/GafferSceneUI/LightTool.cpp:1677
GafferSceneUI::LightTool::preRender (this=0xcb77a10) at src/GafferSceneUI/LightTool.cpp:1837
```

Reply:
Fixed in https://github.com/GafferHQ/gaffer/pull/5218/commits/61072a1f2f11066af3c9126d894dcc6f3bddf7e8